tests/Model/BSONIteratorTest.php

```diff
@@ -0,0 +1,85 @@
+<?php
+
+namespace MongoDB\Tests\Model;
+
+use MongoDB\Exception\UnexpectedValueException;
+use MongoDB\Model\BSONArray;
+use MongoDB\Model\BSONDocument;
+use MongoDB\Model\BSONIterator;
+use MongoDB\Tests\TestCase;
+use stdClass;
+
+class BSONIteratorTest extends TestCase
+{
+    /**
+     * @dataProvider provideTypeMapOptionsAndExpectedDocuments
+     */
+    public function testValidValues(array $typeMap, $binaryString, array $expectedDocuments)
+    {
+        $bsonIt = new BSONIterator($binaryString, $typeMap);
+
+        $this->assertSameDocuments($expectedDocuments, $bsonIt);
+    }
+
+    public function provideTypeMapOptionsAndExpectedDocuments()
+    {
+        return [
+            [
+                [],
+                implode(array_map(
+                    'MongoDB\BSON\fromPHP',
+                    [new stdClass, ['x' => 1]]
+                )),
+                [
+                    new stdClass,
+                    ['x' => 1],
+                ]
+            ],
+            [
+                ['root' => 'array', 'document' => 'array'],
+                implode(array_map(
+                    'MongoDB\BSON\fromPHP',
+                    [new BSONDocument(['foo' => 'bar']), new BSONArray(['a' => 'b'])]
+                )),
+                [
+                    new BSONDocument(['foo' => 'bar']),
+                    new BSONArray(['a' => 'b']),
+                ]
+            ],
+            [
+                [],
+                '',
+                [],
+            ],
+            /*
+                Exception for remaining buffer after 32-bit length field being too small, for
+                both the first and second positions in the sequence.
+            */
+        ];
+    }
+
+    public function testRemainingBufferLength()
+    {
+        $binaryString = '1234';
```

Comment:
For the exception tests, I'd suggest using the base document `fromPHP([])`, which should represent the smallest valid BSON document (in binary form). It contains a 32-bit integer with a value of `5`, followed by a null byte (per the `document` definition in the [spec](http://bsonspec.org/spec.html) where `e_list` is empty).



The first exception is when we don't have sufficient data in the buffer to read the 4-byte length header. For this, I'd like a test for both the very first document and a subsequent document:

 * `testCannotReadLengthFromFirstDocument()`: start with one document as our input, and trim it to the first three bytes (anything less than four will suffice).
 * `testCannotReadLengthFromSubsequentDocument()`: start with two documents as our input, but trim the second document to only three bytes

The second exception is when we _were_ able to read the 4-byte length header, but there is not sufficient data in the buffer to read the remainder.

 * `testCannotReadFirstDocument()`: start with one document as our input and trim to the first four bytes
 * `testCannotReadSecondDocument()`: start with two documents as our input, but trim the second document to only four bytes

In PHPC, I use `pack()` to create these BSON documents. You may decide that's preferable (see: [bson-toJSON_error-002.phpt](https://github.com/mongodb/mongo-php-driver/blob/master/tests/bson/bson-toJSON_error-002.phpt)).